src/Workspaces/Core/Portable/Shared/Extensions/ITypeSymbolExtensions.SubstituteTypesVisitor.cs

```diff
@@ -80,13 +80,15 @@ public override ITypeSymbol VisitNamedType(INamedTypeSymbol symbol)
                     symbol = updatedContainingType.GetTypeMembers(symbol.Name, symbol.Arity).First(m => m.TypeKind == symbol.TypeKind);
                 }
 
-                var substitutedArguments = symbol.TypeArguments.Select(t => t.Accept(this)).ToArray();
-                if (symbol.TypeArguments.SequenceEqual(substitutedArguments))
+                var typeArgumentsWithNullability = symbol.TypeArguments.ZipAsArray(symbol.TypeArgumentsNullableAnnotations, (t, n) => t.WithNullability(n));
+                var substitutedArguments = typeArgumentsWithNullability.Select(t => t.Accept(this));
+                if (typeArgumentsWithNullability.SequenceEqual(substitutedArguments))
                 {
                     return symbol;
                 }
 
-                return _typeGenerator.Construct(symbol.OriginalDefinition, substitutedArguments);
+                // TODO: pass nullability to the substituted arguments once https://github.com/dotnet/roslyn/issues/36046 is fixed
+                return _typeGenerator.Construct(symbol.OriginalDefinition, substitutedArguments.Select(t => t.UnwrapNullabilitySymbol()).ToArray()).WithNullability(symbol.GetNullability());
```

Comment:
>GetNullability [](start = 171, length = 14)

`UnwrapNullabilitySymbol` .... `WithNullability` .... 😭

Reply:
Specifically, does this actually matter if you pass in `TypeSymbolWithNullabelAnnotation`? It seems like it would treat it as if it didn't have nullable information

---
In reply to: [289572503](https://github.com/dotnet/roslyn/pull/36049#discussion_r289572503) [](ancestors = 289572503)

Reply:
Nevermind... this will fail if they ever us `unwrappedType == wrappedType` . We talked about this, and I forgot because it hurts my soul

---
In reply to: [289572711](https://github.com/dotnet/roslyn/pull/36049#discussion_r289572711) [](ancestors = 289572711,289572503)